src/dataflow/render/reduce.rs

```diff
@@ -64,60 +63,118 @@ where
             let keys_clone = group_key.clone();
             let relation_expr_clone = relation_expr.clone();
 
-            let (ok_input, err_input) = self.collection(input).unwrap();
+            // Our first step is to extract `(key, vals)` from `input`.
+            // We do this carefully, attempting to avoid unneccesary allocations
+            // that would result from cloning rows in input arrangements.
+            let group_key_owned = group_key.to_vec();
+            let aggregates_owned = aggregates.to_vec();
+
+            // Tracks the required number of columns to extract.
+            let mut columns_needed = 0;
+            for key in group_key.iter() {
+                for column in key.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+            for aggr in aggregates.iter() {
+                for column in aggr.expr.support() {
+                    columns_needed = std::cmp::max(columns_needed, column + 1);
+                }
+            }
+
+            let mut row_packer = RowPacker::new();
+            let (key_input, mut err_input): (
```

Comment:
I don't understand why this is called `key_input` despite computing both the key _and_ the value.